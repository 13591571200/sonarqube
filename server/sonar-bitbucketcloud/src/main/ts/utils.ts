/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { parse } from 'querystring';
import { resetBundle } from '@sqcore/helpers/l10n';
import l10nbundle from './l10nbundle';
import { WidgetType, AppContext, PullRequestContext } from './types';

declare const AP: {
  getLocation: (callback: (location: string) => void) => void;
  require: Function;
};
const query = parse<{ [key: string]: string }>(window.location.search.replace('?', ''));

export function getAppContext(): AppContext {
  return { jwt: (window as any).bbcJWT || query.jwt || '' };
}

export function getAppKey(): string {
  return (window as any).bbcAppKey || '';
}

export function getDisabled(): boolean {
  return query.disabled !== 'false';
}

export function getWidgetKey(): WidgetType {
  return (window as any).bbcWidgetKey;
}

export function getProjectKey(): string | undefined {
  return (window as any).projectKey;
}

export function getPullRequestContext(): PullRequestContext {
  return {
    prId: (window as any).prId || query.prId,
    repoUuid: getRepoUuid()
  };
}

export function getRepoUuid(): string {
  return query.repoUuid;
}

export function isRepoAdmin(): boolean {
  return query.isRepoAdmin === 'true';
}

export function installLanguageBundle() {
  resetBundle(l10nbundle);
}

export function getRepoSettingsUrl(): Promise<string> {
  return new Promise(resolve =>
    AP.getLocation(location => {
      const parser = document.createElement('a');
      parser.href = location;
      resolve(
        `${parser.origin}/{}/${getRepoUuid()}/admin/addon/admin/${getAppKey()}/repository-config`
      );
    })
  );
}

export function displayDialog(options: { key: string; size?: string }) {
  return getBBCModule('dialog').then(dialog => {
    dialog.create(options);
  });
}

export function displayMessage(
  level: 'error' | 'info' | 'success' | 'warning',
  title: string,
  msg: string,
  options = { closeable: true, fadeout: true, delay: 5000 }
) {
  getBBCModule('messages').then(
    messages => {
      if (messages[level]) {
        messages[level](title, msg, options);
      }
    },
    () => {}
  );
}

export function apiRequest(data: {
  cache: boolean;
  contentType?: string;
  data?: any;
  headers?: any;
  type: string;
  url: string;
}) {
  return getBBCModule('request').then(
    request =>
      new Promise((resolve, reject) =>
        request({
          error: (
            _error: any,
            response: { statusCode: number; body: { errors: Array<{ msg: string }> } }
          ) => {
            reject({ statusCode: response.statusCode, errors: response.body.errors });
          },
          success: resolve,
          ...data
        })
      )
  );
}

export function proxyRequest(data: {
  cache: boolean;
  contentType?: string;
  data?: any;
  headers?: any;
  type: string;
  url: string;
}) {
  return getBBCModule('proxyRequest').then(
    proxyRequest =>
      new Promise((resolve, reject) =>
        proxyRequest({
          error: (
            _error: any,
            response: { statusCode: number; body: { errors: Array<{ msg: string }> } }
          ) => {
            reject({ statusCode: response.statusCode, errors: response.body.errors });
          },
          success: resolve,
          ...data
        })
      )
  );
}

const bbcModule: { [key: string]: any } = {};

export function getBBCModule(key: string) {
  if (bbcModule[key]) {
    return Promise.resolve(bbcModule[key]);
  }
  if (Object.prototype.hasOwnProperty.call(window, 'AP')) {
    return new Promise(resolve =>
      AP.require(key, (moduleRequired: any) => {
        bbcModule[key] = moduleRequired;
        resolve(bbcModule[key]);
      })
    );
  }
  return Promise.reject('Bitbucket cloud AP not defined');
}
