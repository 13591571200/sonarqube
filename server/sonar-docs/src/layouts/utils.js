import { sortBy } from 'lodash';

export function sortNodes(nodes) {
  return nodes.sort((a, b) => {
    if (a.frontmatter.order) {
      return b.frontmatter.order ? a.frontmatter.order - b.frontmatter.order : 1;
    }
    return a.frontmatter.title < b.frontmatter.title ? -1 : 1;
  });
}

const WORDS = 6;

function cutLeadingWords(str) {
  let words = 0;
  for (let i = str.length - 1; i >= 0; i--) {
    if (/\s/.test(str[i])) {
      words++;
    }
    if (words === WORDS) {
      return i > 0 ? `...${str.substring(i + 1)}` : str;
    }
  }
  return str;
}

function cutTrailingWords(str) {
  let words = 0;
  for (let i = 0; i < str.length; i++) {
    if (/\s/.test(str[i])) {
      words++;
    }
    if (words === WORDS) {
      return i < str.length - 1 ? `${str.substring(0, i)}...` : str;
    }
  }
  return str;
}

export function cutWords(tokens) {
  const result = [];
  let length = 0;

  const highlightPos = tokens.findIndex(token => token.marked);
  if (highlightPos > 0) {
    const text = cutLeadingWords(tokens[highlightPos - 1].text);
    result.push({ text, marked: false });
    length += text.length;
  }

  result.push(tokens[highlightPos]);
  length += tokens[highlightPos].text.length;

  for (let i = highlightPos + 1; i < tokens.length; i++) {
    if (length + tokens[i].text.length > 100) {
      const text = cutTrailingWords(tokens[i].text);
      result.push({ text, marked: false });
      return result;
    } else {
      result.push(tokens[i]);
      length += tokens[i].text.length;
    }
  }

  return result;
}

export function highlightMarks(str, marks) {
  const sortedMarks = sortBy(
    [
      ...marks.map(mark => ({ pos: mark.from, start: true })),
      ...marks.map(mark => ({ pos: mark.to, start: false }))
    ],
    mark => mark.pos,
    mark => Number(!mark.start)
  );

  const cuts = [];
  let start = 0;
  let balance = 0;

  for (const mark of sortedMarks) {
    if (mark.start) {
      if (balance === 0 && start !== mark.pos) {
        cuts.push({ text: str.substring(start, mark.pos), marked: false });
        start = mark.pos;
      }
      balance++;
    } else {
      balance--;
      if (balance === 0 && start !== mark.pos) {
        cuts.push({ text: str.substring(start, mark.pos), marked: true });
        start = mark.pos;
      }
    }
  }

  if (start < str.length - 1) {
    cuts.push({ text: str.substr(start), marked: false });
  }

  return cuts;
}
