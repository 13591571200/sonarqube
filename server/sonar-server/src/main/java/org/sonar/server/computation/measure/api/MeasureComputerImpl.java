/*
 * SonarQube, open source software quality management tool.
 * Copyright (C) 2008-2014 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * SonarQube is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * SonarQube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package org.sonar.server.computation.measure.api;

import com.google.common.collect.ImmutableSet;
import java.util.Set;
import org.sonar.api.ce.measure.MeasureComputer;

import static com.google.common.base.Preconditions.checkArgument;

public class MeasureComputerImpl implements MeasureComputer {

  private final Set<String> inputMetricKeys;
  private final Set<String> outputMetrics;
  private final Implementation implementation;

  public MeasureComputerImpl(MeasureComputerBuilderImpl builder) {
    this.inputMetricKeys = ImmutableSet.copyOf(builder.inputMetricKeys);
    this.outputMetrics = ImmutableSet.copyOf(builder.outputMetrics);
    this.implementation = builder.measureComputerImplementation;
  }

  @Override
  public Set<String> getInputMetrics() {
    return inputMetricKeys;
  }

  @Override
  public Set<String> getOutputMetrics() {
    return outputMetrics;
  }

  @Override
  public Implementation getImplementation() {
    return implementation;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }

    MeasureComputerImpl that = (MeasureComputerImpl) o;

    if (!inputMetricKeys.equals(that.inputMetricKeys)) {
      return false;
    }
    return outputMetrics.equals(that.outputMetrics);
  }

  @Override
  public int hashCode() {
    int result = inputMetricKeys.hashCode();
    result = 31 * result + outputMetrics.hashCode();
    return result;
  }

  @Override
  public String toString() {
    return "MeasureComputerImpl{" +
      "inputMetricKeys=" + inputMetricKeys +
      ", outputMetrics=" + outputMetrics +
      ", implementation=" + implementation +
      '}';
  }

  public static class MeasureComputerBuilderImpl implements MeasureComputerBuilder {

    private String[] inputMetricKeys;
    private String[] outputMetrics;
    private Implementation measureComputerImplementation;

    @Override
    public MeasureComputerBuilder setInputMetrics(String... inputMetrics) {
      checkArgument(inputMetrics != null && inputMetrics.length > 0, "At least one input metric must be defined");
      this.inputMetricKeys = inputMetrics;
      return this;
    }

    @Override
    public MeasureComputerBuilder setOutputMetrics(String... outputMetrics) {
      checkArgument(outputMetrics != null && outputMetrics.length > 0, "At least one output metric must be defined");
      this.outputMetrics = outputMetrics;
      return this;
    }

    @Override
    public MeasureComputerBuilder setImplementation(Implementation impl) {
      checkImplementation(impl);
      this.measureComputerImplementation = impl;
      return this;
    }

    @Override
    public MeasureComputer build() {
      checkArgument(this.inputMetricKeys != null, "At least one input metric must be defined");
      checkArgument(this.outputMetrics != null, "At least one output metric must be defined");
      checkImplementation(this.measureComputerImplementation);
      return new MeasureComputerImpl(this);
    }

    private static void checkImplementation(Implementation impl) {
      checkArgument(impl != null, "The implementation is missing");
    }
  }
}
