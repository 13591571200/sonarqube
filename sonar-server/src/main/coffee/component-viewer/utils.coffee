define ->

  splitLongName: (longName) ->
    lastSeparator = longName.lastIndexOf '/'
    dir: longName.substr 0, lastSeparator
    name: longName.substr lastSeparator + 1