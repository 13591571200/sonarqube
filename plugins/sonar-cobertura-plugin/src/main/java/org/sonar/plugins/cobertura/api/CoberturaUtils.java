package org.sonar.plugins.cobertura.api;

import org.sonar.api.CoreProperties;
import org.sonar.api.batch.maven.MavenPlugin;
import org.sonar.api.resources.Project;
import org.sonar.api.utils.Logs;
import org.sonar.plugins.cobertura.CoberturaMavenPluginHandler;

import java.io.File;

/**
 * @since 2.4
 */
public final class CoberturaUtils {

  public static File getReport(Project project) {
    File report = getReportFromProperty(project);
    if (report == null) {
      report = getReportFromPluginConfiguration(project);
    }
    if (report == null) {
      report = getReportFromDefaultPath(project);
    }

    if (report == null || !report.exists() || !report.isFile()) {
      Logs.INFO.warn("Cobertura report not found at {}", report);
      report = null;
    }
    return report;
  }

  private static File getReportFromProperty(Project project) {
    String path = (String) project.getProperty(CoreProperties.COBERTURA_REPORT_PATH_PROPERTY);
    if (path != null) {
      return project.getFileSystem().resolvePath(path);
    }
    return null;
  }

  private static File getReportFromPluginConfiguration(Project project) {
    MavenPlugin mavenPlugin = MavenPlugin.getPlugin(project.getPom(), CoberturaMavenPluginHandler.GROUP_ID, CoberturaMavenPluginHandler.ARTIFACT_ID);
    if (mavenPlugin != null) {
      String path = mavenPlugin.getParameter("outputDirectory");
      if (path != null) {
        return new File(project.getFileSystem().resolvePath(path), "coverage.xml");
      }
    }
    return null;
  }

  private static File getReportFromDefaultPath(Project project) {
    return new File(project.getFileSystem().getReportOutputDir(), "cobertura/coverage.xml");
  }

  private CoberturaUtils() {
  }

}
