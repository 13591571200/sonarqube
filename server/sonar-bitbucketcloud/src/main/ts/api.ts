/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { parseError, parseJSON, request, RequestData } from '@sqcore/helpers/request';
import { Paging, MyProject, IdentityProvider } from '@sqcore/app/types';
import {
  AppContext,
  ProjectData,
  RepositoryData,
  RepositoryMeasure,
  PullRequestData,
  PullRequestContext
} from './types';
import { displayMessage, proxyRequest, apiRequest, getAppKey, getRepoUuid } from './utils';

export function bindProject(data: AppContext & { projectKey: string }): Promise<void | Response> {
  return post('/integration/bitbucketcloud/bind_repo', data).catch(displayWSError);
}

export function displayWSError({ response }: { response: Response }): Promise<Response> {
  parseError({ response }).then(
    msg => {
      displayMessage('error', 'SonarCloud Error', msg);
    },
    () => {}
  );
  return Promise.reject(response);
}

export function getIdentityProviders(): Promise<{ identityProviders: IdentityProvider[] }> {
  return getJSON('/api/users/identity_providers');
}

export function getMyProjects(data: {
  p?: number;
  ps?: number;
}): Promise<{ paging: Paging; projects: MyProject[] }> {
  return getJSON('/api/projects/search_my_projects', data);
}

export function getRepositoryData(data: AppContext): Promise<ProjectData> {
  return getJSON('/integration/bitbucketcloud/repo_widget_data', data).then(
    (repoData: RepositoryData) => {
      const measures: { [key: string]: string } = {};
      repoData.measures.forEach((measure: RepositoryMeasure) => {
        if (measure.value !== undefined) {
          measures[measure.metric] = measure.value;
        }
      });
      return { ...repoData, measures };
    }
  );
}

export function getStoredProperty(property: string) {
  return apiRequest({
    cache: false,
    type: 'GET',
    url: `/2.0/repositories/{}/${getRepoUuid()}/properties/${getAppKey()}/${property}`
  });
}

export function putStoredProperty(property: string, value: Object | string | boolean | number) {
  return apiRequest({
    cache: false,
    contentType: 'application/json',
    data: JSON.stringify(value),
    type: 'PUT',
    url: `/2.0/repositories/{}/${getRepoUuid()}/properties/${getAppKey()}/${property}`
  });
}

export function getPullRequestData(context: PullRequestContext): Promise<PullRequestData> {
  return getProxyJSON(
    `/repositories/${context.repoUuid}/pullrequests/${context.prId}/code_quality`
  );
}

function checkStatus(response: Response): Promise<Response> {
  return new Promise((resolve, reject) => {
    if (response.status >= 200 && response.status < 300) {
      resolve(response);
    } else {
      reject({ response });
    }
  });
}

function getProxyJSON(url: string): Promise<any> {
  return proxyRequest({ cache: false, type: 'GET', url });
}

function getJSON(url: string, data?: RequestData): Promise<any> {
  return request(url)
    .setData(data)
    .submit()
    .then(checkStatus)
    .then(parseJSON);
}

function post(url: string, data?: RequestData): Promise<void> {
  return new Promise((resolve, reject) => {
    request(url)
      .setMethod('POST')
      .setData(data)
      .submit()
      .then(checkStatus)
      .then(() => resolve(), reject);
  });
}
