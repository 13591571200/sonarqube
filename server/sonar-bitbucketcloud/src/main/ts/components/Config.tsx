/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import Button from '@atlaskit/button';
import { CheckboxStateless } from '@atlaskit/checkbox';
import Select, { createFilter } from '@atlaskit/select';
import Spinner from '@atlaskit/spinner';
import { getBaseUrl } from '@sqcore/helpers/urls';
import HelpLink from './HelpLink';
import LoginForm from './LoginForm';
import SonarCloudIcon from './SonarCloudIcon';
import { AppContext } from '../types';
import { bindProject, displayWSError, getMyProjects, putStoredProperty } from '../api';
import { displayMessage } from '../utils';

interface ProjectOption {
  label: string;
  value: string;
}

interface Props {
  context: AppContext;
  disabled: boolean;
  projectKey?: string;
  updateDisabled: (disabled: boolean) => void;
  updateProjectKey: (projectKey: string) => void;
}

interface State {
  authenticated: boolean;
  saving: boolean;
  bindSuccess?: boolean;
  disabled: boolean;
  loading: boolean;
  projects: ProjectOption[];
  selectedProject?: ProjectOption;
}

export default class Config extends React.PureComponent<Props, State> {
  filterOption = createFilter({
    ignoreCase: true,
    ignoreAccents: true,
    stringify: (option: ProjectOption) => `${option.label} ${option.value}`,
    trim: true,
    matchFrom: 'any'
  });

  mounted = false;

  constructor(props: Props) {
    super(props);
    this.state = {
      authenticated: false,
      saving: false,
      disabled: props.disabled,
      loading: true,
      projects: []
    };
  }

  componentDidMount() {
    this.mounted = true;
    this.fetchMyProjects();
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  get selectedProject() {
    return (
      this.state.selectedProject || this.state.projects.find(p => p.value === this.props.projectKey)
    );
  }

  fetchMyProjects = () => {
    getMyProjects({ ps: 500 }).then(
      ({ projects }) => {
        if (this.mounted) {
          const projectOptions = projects.map(p => ({ label: p.name, value: p.key }));
          this.setState({
            authenticated: true,
            loading: false,
            projects: projectOptions
          });
        }
      },
      ({ response }) => {
        if (this.mounted && response && response.status === 401) {
          this.setState({ authenticated: false, loading: false });
        } else {
          displayWSError({ response });
        }
      }
    );
  };

  handleDisabledChange = () => {
    this.setState(state => ({ disabled: !state.disabled }));
  };

  handleReload = () => {
    window.location.reload();
  };

  handleChange = (selectedProject?: ProjectOption) => {
    if (selectedProject && !Array.isArray(selectedProject)) {
      this.setState({ selectedProject });
    }
  };

  handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const { disabled } = this.state;

    let updateBinding = false;
    const promises: Promise<any>[] = [];
    if (disabled !== this.props.disabled) {
      promises.push(
        putStoredProperty('disabled', disabled).then(() => {
          this.props.updateDisabled(disabled);
        })
      );
    }

    const { selectedProject } = this;
    if (selectedProject && selectedProject.value !== this.props.projectKey) {
      updateBinding = true;
      promises.push(
        bindProject({ ...this.props.context, projectKey: selectedProject.value }).then(() => {
          this.props.updateProjectKey(selectedProject.value);
        })
      );
    }
    if (promises.length > 0) {
      (document.activeElement as HTMLElement).blur();
      this.setState({ saving: true });
      Promise.all(promises).then(
        () => {
          if (this.mounted) {
            displayMessage(
              'success',
              'SonarCloud',
              updateBinding
                ? 'SonarCloud project successfully linked, check the new Code Quality widget on your repository overview.'
                : 'Widget visibility successfully changed.'
            );
            this.setState({ saving: false });
          }
        },
        () => {
          if (this.mounted) {
            this.setState({ saving: false });
          }
        }
      );
    }
  };

  renderContainer = (children: React.ReactNode) => {
    return (
      <>
        <h2>SonarCloud Settings</h2>
        <div className="settings-content">
          <div className="settings-logo">
            <SonarCloudIcon size={128} />
          </div>
          <p className="settings-description">
            To display the quality of your repository, you have to link it with a project analyzed
            on SonarCloud.
          </p>
          {children}
        </div>
      </>
    );
  };

  renderProjectsSelect = () => {
    const { projects } = this.state;

    if (!projects || projects.length <= 0) {
      return (
        <>
          <p>
            You don&apos;t have any project on SonarCloud yet:{' '}
            <a href={getBaseUrl() + '/onboarding'} rel="noopener noreferrer" target="_blank">
              Analyse a project
            </a>
          </p>
        </>
      );
    }

    return (
      <div className="settings-projects">
        <label htmlFor="projects-select">Project to link to</label>
        <Select
          autoFocus={true}
          className="settings-projects-select"
          filterOption={this.filterOption}
          id="projects-select"
          isClearable={false}
          isSearchable={true}
          maxMenuHeight={300}
          onChange={this.handleChange}
          options={projects}
          placeholder="Select a project"
          value={this.selectedProject}
        />
        <small>You see only the projects you administer.</small>
      </div>
    );
  };

  render() {
    const { authenticated, disabled, loading } = this.state;

    if (loading) {
      return this.renderContainer(
        <div className="huge-spacer-top">
          <Spinner size="large" />
        </div>
      );
    }

    const { selectedProject } = this;
    const hasChanged =
      (selectedProject && selectedProject.value !== this.props.projectKey) ||
      this.props.disabled !== disabled;

    return this.renderContainer(
      <>
        {!authenticated && <LoginForm onReload={this.handleReload} />}
        <form className="settings-form" onSubmit={this.handleSubmit}>
          {authenticated && this.renderProjectsSelect()}
          <div className="display-flex-justify-center">
            <CheckboxStateless
              isChecked={!disabled}
              label="Show repository overview widget"
              name="show-widget"
              onChange={this.handleDisabledChange}
              value="show-widget"
            />
          </div>
          <div className="ak-field-group">
            <Button
              appearance="primary"
              isDisabled={this.state.saving || !hasChanged}
              type="submit">
              Save
            </Button>
          </div>
        </form>
        <HelpLink />
      </>
    );
  }
}
