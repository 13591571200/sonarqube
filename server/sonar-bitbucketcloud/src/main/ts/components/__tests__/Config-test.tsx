/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import Config from '../Config';
import { bindProject, getMyProjects } from '../../api';
import { isManualBindingAllowed } from '../../utils';

jest.mock('../../utils', () => ({
  displayMessage: jest.fn(),
  isManualBindingAllowed: jest.fn(() => true)
}));

jest.mock('../../api', () => ({
  bindProject: jest.fn(() => Promise.resolve()),
  displayWSError: jest.fn(),
  getMyProjects: jest.fn(() =>
    Promise.resolve({
      paging: { pageIndex: 1, pageSize: 10, total: 3 },
      projects: [
        { key: 'foo', links: [], name: 'Foo' },
        { key: 'bar', links: [], name: 'Bar' },
        { key: 'baz', links: [], name: 'FooBar' }
      ]
    })
  )
}));

const CONTEXT = { jwt: '' };

beforeEach(() => {
  (bindProject as jest.Mock<any>).mockClear();
  (getMyProjects as jest.Mock<any>).mockClear();
});

it('should display correctly', async () => {
  const wrapper = getWrapper();
  expect(wrapper).toMatchSnapshot();

  await new Promise(setImmediate);
  wrapper.update();
  expect(getMyProjects).toHaveBeenCalled();
  expect(wrapper).toMatchSnapshot();
});

it('should display correctly for auto binding', () => {
  (isManualBindingAllowed as jest.Mock<any>).mockReturnValue(false);
  expect(getWrapper()).toMatchSnapshot();
  (isManualBindingAllowed as jest.Mock<any>).mockReturnValue(true);
});

it('should display correctly for auto binding with already a projectKey', () => {
  (isManualBindingAllowed as jest.Mock<any>).mockReturnValue(false);
  expect(getWrapper()).toMatchSnapshot();
  (isManualBindingAllowed as jest.Mock<any>).mockReturnValue(true);
});

it('should display the authentication component and the display checkbox', async () => {
  (getMyProjects as jest.Mock<any>).mockImplementationOnce(() =>
    Promise.reject({ response: { status: 401 } })
  );
  const wrapper = getWrapper();
  await new Promise(setImmediate);
  wrapper.update();
  expect(wrapper.find('LoginForm').exists()).toBeTruthy();
  expect(wrapper.find('.settings-form')).toMatchSnapshot();
});

it('should correctly handle select interactions', async () => {
  const wrapper = getWrapper({ projectKey: undefined });
  await new Promise(setImmediate);
  wrapper.update();
  expect(wrapper.find('WithAnalyticsContext').prop('isDisabled')).toBeTruthy();

  // Check the select event
  const SelectWrapper = wrapper.find('AtlaskitSelect');
  const projectOption = { label: 'Bar', value: 'bar' };
  (SelectWrapper.prop('onChange') as Function)(projectOption);
  wrapper.update();
  expect(wrapper.state('selectedProject')).toEqual(projectOption);
  expect(wrapper.find('WithAnalyticsContext').prop('isDisabled')).toBeFalsy();
});

it('should correctly bind a project', async () => {
  const updateProjectKey = jest.fn();
  const wrapper = getWrapper({ projectKey: undefined, updateProjectKey });
  await new Promise(setImmediate);
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('WithAnalyticsContext').prop('isDisabled')).toBeTruthy();

  (wrapper.find('AtlaskitSelect').prop('onChange') as Function)({ label: 'Baz', value: 'baz' });
  wrapper.update();
  expect(wrapper.find('WithAnalyticsContext').prop('isDisabled')).toBeFalsy();
  wrapper.find('form').simulate('submit', { preventDefault: () => {} });
  expect(wrapper.find('WithAnalyticsContext').prop('isDisabled')).toBeTruthy();
  expect(bindProject).toHaveBeenCalledWith({ ...CONTEXT, projectKey: 'baz' });

  await new Promise(setImmediate);
  expect(updateProjectKey).toHaveBeenCalledWith('baz');
});

function getWrapper(props = {}) {
  return shallow(
    <Config
      context={CONTEXT}
      disabled={false}
      projectKey="foo"
      updateDisabled={jest.fn()}
      updateProjectKey={jest.fn()}
      {...props}
    />
  );
}
