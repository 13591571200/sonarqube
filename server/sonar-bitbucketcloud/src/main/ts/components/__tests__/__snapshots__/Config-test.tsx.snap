// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly bind a project 1`] = `
<React.Fragment>
  <h2>
    SonarCloud Settings
  </h2>
  <div
    className="settings-content"
  >
    <div
      className="settings-logo"
    >
      <SonarCloudIcon
        size={128}
      />
    </div>
    <p
      className="settings-description"
    >
      To display the quality of your repository, you have to link it with a project analyzed on SonarCloud.
    </p>
    <React.Fragment>
      <form
        className="settings-form"
        onSubmit={[Function]}
      >
        <div
          className="settings-projects"
        >
          <label
            htmlFor="projects-select"
          >
            Project to link to
          </label>
          <AtlaskitSelect
            autoFocus={true}
            className="settings-projects-select"
            filterOption={[Function]}
            id="projects-select"
            isClearable={false}
            isSearchable={true}
            maxMenuHeight={300}
            onChange={[Function]}
            onClickPreventDefault={true}
            options={
              Array [
                Object {
                  "label": "Foo",
                  "value": "foo",
                },
                Object {
                  "label": "Bar",
                  "value": "bar",
                },
                Object {
                  "label": "FooBar",
                  "value": "baz",
                },
              ]
            }
            placeholder="Select a project"
            spacing="default"
            validationState="default"
          />
          <small>
            You see only the projects you administer.
          </small>
        </div>
        <div
          className="display-flex-justify-center"
        >
          <CheckboxStateless
            isChecked={true}
            label="Show repository overview widget"
            name="show-widget"
            onChange={[Function]}
            value="show-widget"
          />
        </div>
        <div
          className="ak-field-group"
        >
          <WithAnalyticsContext(WithAnalyticsEvents(WithDeprecationWarnings(Button)))
            analyticsContext={Object {}}
            appearance="primary"
            isDisabled={true}
            type="submit"
          >
            Save
          </WithAnalyticsContext(WithAnalyticsEvents(WithDeprecationWarnings(Button)))>
        </div>
      </form>
      <HelpLink />
    </React.Fragment>
  </div>
</React.Fragment>
`;

exports[`should display correctly 1`] = `
<React.Fragment>
  <h2>
    SonarCloud Settings
  </h2>
  <div
    className="settings-content"
  >
    <div
      className="settings-logo"
    >
      <SonarCloudIcon
        size={128}
      />
    </div>
    <p
      className="settings-description"
    >
      To display the quality of your repository, you have to link it with a project analyzed on SonarCloud.
    </p>
    <div
      className="huge-spacer-top"
    >
      <Spinner
        delay={100}
        invertColor={false}
        isCompleting={false}
        onComplete={[Function]}
        size="large"
      />
    </div>
  </div>
</React.Fragment>
`;

exports[`should display correctly 2`] = `
<React.Fragment>
  <h2>
    SonarCloud Settings
  </h2>
  <div
    className="settings-content"
  >
    <div
      className="settings-logo"
    >
      <SonarCloudIcon
        size={128}
      />
    </div>
    <p
      className="settings-description"
    >
      To display the quality of your repository, you have to link it with a project analyzed on SonarCloud.
    </p>
    <React.Fragment>
      <form
        className="settings-form"
        onSubmit={[Function]}
      >
        <div
          className="settings-projects"
        >
          <label
            htmlFor="projects-select"
          >
            Project to link to
          </label>
          <AtlaskitSelect
            autoFocus={true}
            className="settings-projects-select"
            filterOption={[Function]}
            id="projects-select"
            isClearable={false}
            isSearchable={true}
            maxMenuHeight={300}
            onChange={[Function]}
            onClickPreventDefault={true}
            options={
              Array [
                Object {
                  "label": "Foo",
                  "value": "foo",
                },
                Object {
                  "label": "Bar",
                  "value": "bar",
                },
                Object {
                  "label": "FooBar",
                  "value": "baz",
                },
              ]
            }
            placeholder="Select a project"
            spacing="default"
            validationState="default"
            value={
              Object {
                "label": "Foo",
                "value": "foo",
              }
            }
          />
          <small>
            You see only the projects you administer.
          </small>
        </div>
        <div
          className="display-flex-justify-center"
        >
          <CheckboxStateless
            isChecked={true}
            label="Show repository overview widget"
            name="show-widget"
            onChange={[Function]}
            value="show-widget"
          />
        </div>
        <div
          className="ak-field-group"
        >
          <WithAnalyticsContext(WithAnalyticsEvents(WithDeprecationWarnings(Button)))
            analyticsContext={Object {}}
            appearance="primary"
            isDisabled={true}
            type="submit"
          >
            Save
          </WithAnalyticsContext(WithAnalyticsEvents(WithDeprecationWarnings(Button)))>
        </div>
      </form>
      <HelpLink />
    </React.Fragment>
  </div>
</React.Fragment>
`;

exports[`should display the authentication component and the display checkbox 1`] = `
<form
  className="settings-form"
  onSubmit={[Function]}
>
  <div
    className="display-flex-justify-center"
  >
    <CheckboxStateless
      isChecked={true}
      label="Show repository overview widget"
      name="show-widget"
      onChange={[Function]}
      value="show-widget"
    />
  </div>
  <div
    className="ak-field-group"
  >
    <WithAnalyticsContext(WithAnalyticsEvents(WithDeprecationWarnings(Button)))
      analyticsContext={Object {}}
      appearance="primary"
      isDisabled={true}
      type="submit"
    >
      Save
    </WithAnalyticsContext(WithAnalyticsEvents(WithDeprecationWarnings(Button)))>
  </div>
</form>
`;
